hail/hail/ir-gen/src/Main.scala

```diff
@@ -0,0 +1,1239 @@
+        s"override def childrenSeq: IndexedSeq[BaseIR] = IndexedSeq.empty"
+      else
+        s"override lazy val childrenSeq: IndexedSeq[BaseIR] = $children"
+      val extraMethods =
+        this.extraMethods :+ childrenSeqDef :+ copyMethod
+      val constraints = this.constraints ++ attsAndChildren.flatMap(_.constraints)
+      if (constraints.nonEmpty || extraMethods.nonEmpty) {
+        (
+          " {" +
+            (if (constraints.nonEmpty)
+               constraints.map(c => s"  require($c)").mkString("\n", "\n", "\n")
+             else "")
+            + (
+              if (extraMethods.nonEmpty)
+                extraMethods.map("  " + _).mkString("\n", "\n", "\n")
+              else ""
+            )
+            + "}"
+        )
+      } else ""
+    }
+
+    private def classDef =
+      (if (docstring.nonEmpty) s"\n/** $docstring*/\n" else "") + classDecl + classBody
+
+    private def companionDef = if (hasCompanionExtension)
+      s"object $name extends ${name}CompanionExt\n"
+    else ""
+
+    def generateDef: String = companionDef + classDef + "\n"
+  }
+}
+
+object Main {
+  val irdsl: IRDSL = IRDSL_Impl
+  import irdsl._
+
+  private val errorID = in("errorID", att("Int")).withDefault("ErrorIDs.NO_ERROR")
+
+  private def _typ(t: String = "Type") = in("_typ", att(t))
+
+  private val mmPerElt =
+    in("requiresMemoryManagementPerElement", att("Boolean")).withDefault("false")
+
+  private def allNodes: Seq[NodeDef] = {
+
+    val r = Seq.newBuilder[NodeDef]
+
+    r += node("I32", in("x", att("Int"))).withTraits(TrivialIR)
+    r += node("I64", in("x", att("Long"))).withTraits(TrivialIR)
+    r += node("F32", in("x", att("Float"))).withTraits(TrivialIR)
+    r += node("F64", in("x", att("Double"))).withTraits(TrivialIR)
+    r += node("Str", in("x", att("String"))).withTraits(TrivialIR)
+      .withMethod(
+        "override def toString(): String = s\"\"\"Str(\"${StringEscapeUtils.escapeString(x)}\")\"\"\""
+      ): @nowarn("msg=possible missing interpolator")
+    r += node("True").withTraits(TrivialIR)
+    r += node("False").withTraits(TrivialIR)
+    r += node("Void").withTraits(TrivialIR)
+    r += node("NA", _typ()).withTraits(TrivialIR)
+    r += node("UUID4", in("id", att("String")))
+      .withDocstring(
+        """WARNING! This node can only be used when trying to append a one-off,
+          |random string that will not be reused elsewhere in the pipeline.
+          |Any other uses will need to write and then read again; this node is non-deterministic
+          |and will not e.g. exhibit the correct semantics when self-joining on streams.
+          |""".stripMargin
+      )
+      .withCompanionExtension
+
+    r += node(
+      "Literal",
+      in("_typ", att("Type")).withConstraint(self => s"!CanEmit($self)"),
+      in("value", att("Annotation")).withConstraint(self => s"$self != null"),
+    )
+      .withCompanionExtension
+      .withMethod(
+        """// expensive, for debugging
```

Comment:
This has given me the idea to write a macro that includes this code if we're building in debug mode - would be nicer than having to find and uncomment all this stuff!